# UTS PENAMBANGAN DATA

**Nama  : Taufiqu Reza Yoga Pratama**

**Nim   : 210411100014**

**Kelas : Pendata_C**

**1. Metode KNN**

Klasifikasi data KNN adalah salah satu aplikasi dari metode K-Nearest Neighbors (KNN) dalam pembelajaran mesin. Klasifikasi data KNN mengambil sekelompok data dengan fitur yang mirip dan mengklasifikasikannya ke dalam kelas atau label yang sama.

Dalam KNN, setiap data pada dasarnya diwakili sebagai sebuah titik dalam ruang fitur multidimensi. Ketika KNN melakukan klasifikasi data baru, ia akan mencari sekelompok tetangga terdekat dari data baru tersebut dalam ruang fitur tersebut. KNN kemudian mengambil mayoritas label dari tetangga terdekat sebagai label yang akan diberikan pada data baru tersebut. Oleh karena itu, KNN disebut sebagai metode "lazy learning" karena tidak memerlukan proses pelatihan yang kompleks seperti halnya dengan metode pembelajaran mesin yang lainnya.

**Langkah 1**

Import library yang diperlukan (pandas, numpy, sklearn)

In [49]:
import pandas as pd
import numpy as np
from sklearn.neighbors import KNeighborsRegressor 
from sklearn.preprocessing import LabelEncoder

**Langkah 2**

Import data yang ingin kita analisa

In [50]:
df = pd.read_csv('https://gist.githubusercontent.com/Rezayoga1383/4a5ce7439abb0e49c53098c6fe61c263/raw/a2f25a0e53cf8c560cebd703e740ba7918a9e2e8/uts.csv')

table_row = df.shape[0]
df.head(table_row)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

**Langkah 3**

Melakukan pengecekan apakah ada data yang kosong di dalam file data yang ingin kita analisa

In [51]:
df_copy = df.copy()
df_copy.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

**Langkah 4**

Melakukan normalisasi data, di normalisasi data ini kita mengubah nilai yang ada di dalam fitur ke dalam bentuk bilangan biner atau rentan 0 hingga 1. 

Disini kita menggunakan metode Min-Max, di bawah ini adalah rumus dari metode Min-Max : 

$$
\text { Min }=(L \times R j)+S S \operatorname{Max}=2(L \times R j) \text { Min-Max }
$$

yang perlu digaris bawahi disini pada fitur target tidak akan dinormalisasi.

In [52]:
target_column = df_copy['target']
df_copy = df_copy[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']].copy()
df_copy = (df_copy - df_copy.min())/(df_copy.max() - df_copy.min())
df_copy['target'] = target_column.copy()
df_copy.head(table_row)

age  sex        cp  trestbps      chol  fbs  restecg   thalach  \
0    0.708333  1.0  1.000000  0.481132  0.244292  1.0      0.0  0.603053   
1    0.166667  1.0  0.666667  0.339623  0.283105  0.0      0.5  0.885496   
2    0.250000  0.0  0.333333  0.339623  0.178082  0.0      0.0  0.770992   
3    0.562500  1.0  0.333333  0.245283  0.251142  0.0      0.5  0.816794   
4    0.583333  0.0  0.000000  0.245283  0.520548  0.0      0.5  0.702290   
..        ...  ...       ...       ...       ...  ...      ...       ...   
298  0.583333  0.0  0.000000  0.433962  0.262557  0.0      0.5  0.396947   
299  0.333333  1.0  1.000000  0.150943  0.315068  0.0      0.5  0.465649   
300  0.812500  1.0  0.000000  0.471698  0.152968  1.0      0.5  0.534351   
301  0.583333  1.0  0.000000  0.339623  0.011416  0.0      0.5  0.335878   
302  0.583333  0.0  0.333333  0.339623  0.251142  0.0      0.0  0.786260   

     exang   oldpeak  slope    ca      thal  target  
0      0.0  0.370968    0.0  0.00  0.333333       1  
1      0.0  0.564516    0.0  0.00  0.666667       1  
2      0.0  0.225806    1.0  0.00  0.666667       1  
3      0.0  0.129032    1.0  0.00  0.666667       1  
4      1.0  0.096774    1.0  0.00  0.666667       1  
..     ...       ...    ...   ...       ...     ...  
298    1.0  0.032258    0.5  0.00  1.000000       0  
299    0.0  0.193548    0.5  0.00  1.000000       0  
300    0.0  0.548387    0.5  0.50  1.000000       0  
301    1.0  0.193548    0.5  0.25  1.000000       0  
302    0.0  0.000000    0.5  0.25  0.666667       0  

[303 rows x 14 columns]

**Langkah 5**

1. Langkah pertama adalah menginisialisasi objek knn dari kelas KNeighborsRegressor, yang digunakan untuk melakukan regresi dengan algoritma K-Nearest Neighbors. Selanjutnya, kita melakukan randomisasi indeks dari dataset dengan menggunakan np.random.permutation() dan menetapkan seed 1 agar hasilnya konsisten.

2. Kemudian, dataset dibagi menjadi dua bagian: 75% untuk training dan 25% untuk testing. Data training disimpan dalam variabel train_index dan data testing disimpan dalam variabel testing_data.

3. Setelah itu, melakukan training model menggunakan data training dengan memanggil method fit() pada objek knn. Argumen input berupa fitur-fitur yang digunakan (yaitu age, sex, cp, trestbps, chol, fbs, restecg, thalach, exang, oldpeak, slope, ca, dan thal) dan labelnya adalah target.

4. Selanjutnya, model melakukan prediksi pada data testing dengan memanggil method predict() pada objek knn. Argumen input berupa fitur-fitur dari data testing yang ingin diprediksi.

5. Selanjutnya, akurasi model dihitung dengan mengukur berapa persen data testing yang berhasil diprediksi dengan threshold default 0.5.

6. Terakhir, hasil akurasi model akan dicetak.

In [53]:
knn = KNeighborsRegressor()

# before we split, make sure you randomize the dataset
np.random.seed(1)
index_shuffled = np.random.permutation(df_copy.index)
shuffled_car_data = df_copy.loc[index_shuffled]

# We will use 75% of the data for training and the remaining 25% for testing
train_index = int(shuffled_car_data.shape[0] * 0.75)
training_data = shuffled_car_data.iloc[0:train_index]
testing_data = shuffled_car_data.iloc[train_index:]

# fit the model with the training data
knn.fit(training_data[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']], training_data['target'])

# predict the target of the testing data
prediction = knn.predict(testing_data[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']])

# calculate the accuracy of the model
accuracy = np.mean(np.abs(prediction - testing_data['target']) <= 0.5)

print(f"Akurasi model KNN: {accuracy:.2%}")


Akurasi model KNN: 92.11%


**kesimpulan**

dengan hasil perhitungan di atas kita bisa mendapatkan nilai akurasi dimana dengan menggunakan metode KNN kita mendapatkan nilai akurasi sebesar 92.11%.

**2. Metode Naive Bayes**

**Langkah 1**

Melakukan import library yang kita perlukan dalam menggunakan metode naive bayes (pandas, numpy, dan sklearn)

In [54]:
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn import preprocessing
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder

**Langkah 2**

Melakukan import data yang ingin kita analisa

In [55]:
df = pd.read_csv('https://gist.githubusercontent.com/Rezayoga1383/4a5ce7439abb0e49c53098c6fe61c263/raw/a2f25a0e53cf8c560cebd703e740ba7918a9e2e8/uts.csv')

table_row = df.shape[0]
df.head(table_row)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0     63    1   3       145   233    1        0      150      0      2.3   
1     37    1   2       130   250    0        1      187      0      3.5   
2     41    0   1       130   204    0        0      172      0      1.4   
3     56    1   1       120   236    0        1      178      0      0.8   
4     57    0   0       120   354    0        1      163      1      0.6   
..   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
298   57    0   0       140   241    0        1      123      1      0.2   
299   45    1   3       110   264    0        1      132      0      1.2   
300   68    1   0       144   193    1        1      141      0      3.4   
301   57    1   0       130   131    0        1      115      1      1.2   
302   57    0   1       130   236    0        0      174      0      0.0   

     slope  ca  thal  target  
0        0   0     1       1  
1        0   0     2       1  
2        2   0     2       1  
3        2   0     2       1  
4        2   0     2       1  
..     ...  ..   ...     ...  
298      1   0     3       0  
299      1   0     3       0  
300      1   2     3       0  
301      1   1     3       0  
302      1   1     2       0  

[303 rows x 14 columns]

**Langkah 3**

Melakukan pengecekan apakah ada data yang kosong di dalam file data yang ingin kita analisa

In [56]:
df_copy = df.copy()
df_copy.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

**Langkah 4**

Melakukan pengecekan tipe data yang ada di semua fitur. jika ada tipe data yang tidak bertipe numeric (int, float) maka kita harus rubahnya menjadi tipe data numeric (int, float). 

In [57]:
df_copy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


**Langkah 5**

Melakukan normalisasi data, di normalisasi data ini kita mengubah nilai yang ada di dalam fitur ke dalam bentuk bilangan biner atau rentan 0 hingga 1. 

Disini kita menggunakan metode Min-Max, di bawah ini adalah rumus dari metode Min-Max : 

$$
\text { Min }=(L \times R j)+S S \operatorname{Max}=2(L \times R j) \text { Min-Max }
$$

yang perlu digaris bawahi disini pada fitur target tidak akan dinormalisasi.

In [58]:
target_column = df_copy['target']
df_copy = df_copy[['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach', 'exang', 'oldpeak', 'slope', 'ca', 'thal']].copy()
df_copy = (df_copy - df_copy.min())/(df_copy.max() - df_copy.min())
df_copy['target'] = target_column.copy()
df_copy.head(table_row)

age  sex        cp  trestbps      chol  fbs  restecg   thalach  \
0    0.708333  1.0  1.000000  0.481132  0.244292  1.0      0.0  0.603053   
1    0.166667  1.0  0.666667  0.339623  0.283105  0.0      0.5  0.885496   
2    0.250000  0.0  0.333333  0.339623  0.178082  0.0      0.0  0.770992   
3    0.562500  1.0  0.333333  0.245283  0.251142  0.0      0.5  0.816794   
4    0.583333  0.0  0.000000  0.245283  0.520548  0.0      0.5  0.702290   
..        ...  ...       ...       ...       ...  ...      ...       ...   
298  0.583333  0.0  0.000000  0.433962  0.262557  0.0      0.5  0.396947   
299  0.333333  1.0  1.000000  0.150943  0.315068  0.0      0.5  0.465649   
300  0.812500  1.0  0.000000  0.471698  0.152968  1.0      0.5  0.534351   
301  0.583333  1.0  0.000000  0.339623  0.011416  0.0      0.5  0.335878   
302  0.583333  0.0  0.333333  0.339623  0.251142  0.0      0.0  0.786260   

     exang   oldpeak  slope    ca      thal  target  
0      0.0  0.370968    0.0  0.00  0.333333       1  
1      0.0  0.564516    0.0  0.00  0.666667       1  
2      0.0  0.225806    1.0  0.00  0.666667       1  
3      0.0  0.129032    1.0  0.00  0.666667       1  
4      1.0  0.096774    1.0  0.00  0.666667       1  
..     ...       ...    ...   ...       ...     ...  
298    1.0  0.032258    0.5  0.00  1.000000       0  
299    0.0  0.193548    0.5  0.00  1.000000       0  
300    0.0  0.548387    0.5  0.50  1.000000       0  
301    1.0  0.193548    0.5  0.25  1.000000       0  
302    0.0  0.000000    0.5  0.25  0.666667       0  

[303 rows x 14 columns]

**Langkah 6**

Melakukan pemotongan data (Data Slicing)

1. pisahkan data menjadi 2 yaitu data training dan data test. disini kita menggunakan metode train_test_split dari sklearn.

2. Variabel features_train akan berisi data fitur untuk training, variabel features_test akan berisi data fitur untuk test, variabel target_train akan berisi target untuk training, dan variabel target_test akan berisi target untuk test.

In [59]:
features = df_copy.values[:,:13]  #Variabel features diisi dengan nilai features dari dataset adult_df_rev dengan slicing [:,:13] artinya mengambil seluruh baris dan kolom 0 hingga kolom 3 (kolom ke-13 tidak diambil karena merupakan variabel target).
target = df_copy.values[:,13]  #Variabel target diisi dengan nilai target dari dataset adult_df_rev dengan slicing [:,13] artinya mengambil seluruh baris dan hanya kolom ke-13 (variabel target).
features_train, features_test, target_train, target_test = train_test_split(features,target, test_size = 0.33, random_state = 10)  #Kemudian menggunakan fungsi train_test_split, dataset yang sudah dipisahkan dijadikan argumen, dengan tambahan argumen test_size = 0.33 yang berarti dataset akan dibagi menjadi 33% data uji dan 67% data latih, dan random_state = 10 sebagai seed untuk menghasilkan hasil yang sama setiap kali dijalankan.

**Langkah 7**

Menerapkan Gausian NaiveBayes dengan cara menggunakan modul GaussianNB dari sklearn.

In [60]:
clf = GaussianNB()
clf.fit(features_train, target_train)
target_pred = clf.predict(features_test)

**Langkah 8**

Nilai akurasi model Naive Bayes

In [61]:
accuracy = accuracy_score(target_test, target_pred, normalize = True)

print(f"Akurasi Model NaiveBayes: {accuracy:.2%}")

Akurasi Model NaiveBayes: 79.00%


**Kesimpulan**

dengan hasil perhitungan di atas kita bisa mendapatkan nilai akurasi dimana dengan menggunakan metode NaiveBayes kita mendapatkan nilai akurasi sebesar 79.00%.